### ISSUE TITLE
YOLOv11 Segmentation Model Performance Degrades Significantly After ONNX Conversion (Missing Objects in C++ Inference)

### ISSUE THREAD

**Issue by happybear1015** [ultralytics/ultralytics#21673]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Export

### Bug

The YOLOv11 segmentation model performs well in PyTorch (.pt) format, but its performance significantly deteriorates after conversion to ONNX. Python predictions yield accurate results (all objects in an image are segmented correctly), while C++ ONNX inference produces poor results (some objects in the same image are not segmented). Different export parameters (dynamic/half/simplify) have been tried, but the issue persists.

**Training code:**  
```python
from ultralytics import YOLO  
# Load a model  
model = YOLO("yolo11m-seg.pt")    
# Train the model  
results = model.train(data="coco8-seg.yaml", epochs=100, imgsz=1280, device=[0, 1])  
```

**Python inference code:**  
```python
from ultralytics import YOLO  
# Load a model  
model = YOLO("./runs/segment/train/weights/best.pt")  # pretrained YOLO11n model  

# Run batched inference on a list of images  
image1 = "datasets/PF-SPEED/images/train/ym.jpg"  
results = model([image1])  # return a list of Results objects  

# Process results list  
for result in results:  
    boxes = result.boxes  # Boxes object for bounding box outputs  
    masks = result.masks  # Masks object for segmentation masks outputs  
    keypoints = result.keypoints  # Keypoints object for pose outputs  
    probs = result.probs  # Probs object for classification outputs  
    obb = result.obb  # Oriented boxes object for OBB outputs  
    result.show()  # display to screen  
    result.save(filename="result.jpg")  # save to disk  
```

**ONNX conversion code:**  
```python
from ultralytics import YOLO  
model = YOLO("./runs/segment/train/weights/best.pt")  
# success = model.export(format="onnx", dynamic=True, simplify=True)
```

**C++ prediction code:**

```cpp
#include <opencv2/highgui/highgui.hpp>
#include <iostream>
#include <string>
#include <fstream>
#include <filesystem> // For directory operations
#include <chrono>     // For high-precision timing
#include "YOLO11Seg.hpp"
#include <numeric>    // For calculating averages

namespace fs = std::filesystem; // Simplify namespace

// Get all image files in a directory
std::vector<std::string> getImageFiles(const std::string& directory) {
    std::vector<std::string> imageFiles;
    try {
        for (const auto& entry : fs::directory_iterator(directory)) {
            if (entry.is_regular_file()) {
                std::string filename = entry.path().string();
                // Check file extension
                if (filename.find(".jpg") != std::string::npos ||
                    filename.find(".png") != std::string::npos ||
                    filename.find(".jpeg") != std::string::npos) {
                    imageFiles.push_back(filename);
                }
            }
        }
    }
    catch (const fs::filesystem_error& e) {
        std::cerr << "Failed to open directory: " << directory << ", error: " << e.what() << std::endl;
    }
    return imageFiles;
}

int main()
{
    // Create results directory
    const std::string resultsDir = "results";
    if (!fs::exists(resultsDir)) {
        fs::create_directory(resultsDir);
    }

    // Paths for model, labels, and test images
    const std::string labelsPath = "../models/coco.names";
    const std::string imageDirectory = "../data"; // Image directory path
    const std::string modelPath = "../models/best.onnx";

    bool isGPU = false; // false means using CPU; true means using GPU
    // Initialize YOLO detector
    YOLOv11SegDetector detector(modelPath, labelsPath, isGPU);

    // Get all image files in directory
    std::vector<std::string> imageFiles = getImageFiles(imageDirectory);
    if (imageFiles.empty()) {
        std::cerr << "Error: No image files in directory!\n";
        return -1;
    }

    // Create timing log file
    std::ofstream timeFile(resultsDir + "/inference_times.txt");
    if (!timeFile.is_open()) {
        std::cerr << "Failed to create timing log file!\n";
        return -1;
    }

    // Store all inference times
    std::vector<long long> inferenceTimes;

    // Process each image in directory
    for (const auto& imagePath : imageFiles) {
        // Load image
        cv::Mat image = cv::imread(imagePath);
        if (image.empty()) {
            std::cerr << "Error: Failed to open or find image: " << imagePath << "\n";
            continue;
        }

        // Start pure inference timing
        auto inferenceStart = std::chrono::high_resolution_clock::now();
        std::vector<Segmentation> results = detector.segment(image, 0.0f, 0.05f);
        auto inferenceEnd = std::chrono::high_resolution_clock::now();
        auto inferenceTime = std::chrono::duration_cast<std::chrono::milliseconds>(inferenceEnd - inferenceStart);

        // Record inference time
        inferenceTimes.push_back(inferenceTime.count());
        timeFile << "Image: " << fs::path(imagePath).filename().string()
            << " Inference time: " << inferenceTime.count() << " ms\n";

        std::cout << "Instance segmentation completed, pure inference time: " << inferenceTime.count() << " ms" << std::endl;
        std::cout << "Number of detected objects: " << results.size() << std::endl;

        // Get image name (without path and extension)
        std::string imageName = fs::path(imagePath).stem().string();

        // Process each detection result (this part is not timed)
        for (size_t i = 0; i < results.size(); ++i) {
            const auto& seg = results[i];
            std::cout << "Class ID: " << seg.classId << ", Confidence: " << seg.conf << std::endl;
            std::cout << "Bounding box: (" << seg.box.x << ", " << seg.box.y << ")"
                << " Width: " << seg.box.width << " Height: " << seg.box.height << std::endl;

            // Condition: Only save mask coordinates for qualified objects
            bool isTargetObject = (seg.conf > 0.0); // Example condition: class is 0 and confidence > 0.5
            if (isTargetObject && !seg.mask.empty()) {
                // Extract mask point coordinates
                std::vector<cv::Point> maskPoints;
                cv::findNonZero(seg.mask, maskPoints); // Get coordinates of all non-zero pixels in mask

                // Write mask coordinates to file with image name and index
                std::string filename = resultsDir + "/mask_points_" + imageName + "_" + std::to_string(i) + ".txt";
                std::ofstream outFile(filename);
                if (outFile.is_open()) {
                    for (const auto& point : maskPoints) {
                        outFile << point.x << " " << point.y << "\n"; // Write coordinates
                    }
                    outFile.close();
                    std::cout << "Mask coordinates written to file: " << filename << std::endl;
                }
                else {
                    std::cerr << "Failed to open file: " << filename << std::endl;
                }
            }
            else {
                std::cout << "Object " << i << " doesn't meet requirements, skipping save." << std::endl;
            }
        }

        // Draw results (this part is not timed)
        detector.drawSegmentationsAndBoxes(image, results); // Draw masks and bounding boxes

        // Save image to file
        std::string resultImagePath = resultsDir + "/result_" + fs::path(imagePath).filename().string();
        cv::imwrite(resultImagePath, image);
    }

    // Calculate and record total and average time
    if (!inferenceTimes.empty()) {
        long long totalTime = std::accumulate(inferenceTimes.begin(), inferenceTimes.end(), 0LL);
        double averageTime = static_cast<double>(totalTime) / inferenceTimes.size();

        timeFile << "\nTotal inference time: " << totalTime << " ms\n";
        timeFile << "Average inference time: " << averageTime << " ms\n";

        std::cout << "\nProcessing completed!\n";
        std::cout << "Total inference time: " << totalTime << " ms\n";
        std::cout << "Average inference time: " << averageTime << " ms\n";
    }

    timeFile.close();
    return 0;
}
```

### Minimal Reproducible Example

AS show！

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @happybear1015, thanks for opening this issue and for using Ultralytics 🚀! This is an automated response to help get things moving quickly. An Ultralytics engineer will review your report and assist you soon.

Please check the [Docs](https://docs.ultralytics.com/) for helpful guides, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples that often cover common questions.

Since this looks like a 🐛 Bug Report affecting ONNX export and C++ inference, could you please provide a complete [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE)? This should include:
-Exact export command used (all args) and the resulting ONNX opset shown in logs  
-Sample input image and its preprocessing used in C++ (resize/letterbox, normalization, channel order)  
-ONNX Runtime or other backend version, build flags, and CPU/GPU choice  
-Postprocessing details in C++ (NMS thresholds, mask proto handling, any transforms)  
-Model files and shapes: the exported .onnx and names.yaml if possible

If this is a custom training ❓ question instead, please add dataset samples, training logs, and confirm you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). 🙂

Join the Ultralytics community where it suits you best:
-For real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧  
-For discussions: [Discourse](https://community.ultralytics.com/)  
-For threads: [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

Please upgrade to the latest ultralytics and verify all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a Python>=3.8 environment with PyTorch>=1.8 to ensure the issue isn’t already resolved:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

-Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
-Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
-Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
-Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
Is the issue present in ONNX inference with Ultralytics?

**Comment by happybear1015** [comment]
> Is the issue present in ONNX inference with Ultralytics?

**I tried the following code, and the results are good! Why the C++ infer result not good？**

```
from ultralytics import YOLO
# Load a model
model = YOLO("./runs/segment/train/weights/best.onnx")  

image1="datasets/images/train/ym.jpg"
results = model([image1])


for result in results:
    boxes = result.boxes 
    masks = result.masks 
    keypoints = result.keypoints 
    probs = result.probs
    obb = result.obb 
    result.show() 
    result.save(filename="result.jpg") 
```

**Comment by happybear1015** [comment]
I soved this prolbem！

**Comment by glenn-jocher** [comment]
Great to hear it’s resolved—could you share the root cause and fix for others, and close the issue if everything looks good?

**Comment by glenn-jocher** [comment]
@zrg0712 This usually comes from a C++ post-processing mismatch: your NMS IoU of 0.05 will over‑suppress boxes—match Ultralytics defaults (`conf=0.25`, `iou=0.7`) or bake NMS into the graph when exporting, then ensure preprocessing is identical (letterbox resize to `imgsz`, RGB, 1/255) and that you use both ONNX outputs (`output0` boxes/scores/mask coeffs, `output1` proto masks). For a quick fix, re‑export with NMS and dynamic shapes:
```python
from ultralytics import YOLO
YOLO("runs/segment/train/weights/best.pt").export(format="onnx", nms=True, dynamic=True, simplify=True)
```
Details are in the ONNX export guide and the `export_onnx` reference (outputs and args): see [ONNX Export for YOLO11 Models](https://docs.ultralytics.com/integrations/onnx/) and [export_onnx API reference](https://docs.ultralytics.com/reference/engine/exporter/#export_onnx). If results still differ, please open a new issue with a minimal C++ snippet showing your preprocessing and NMS settings for visibility.
